### ISSUE TITLE
Some smw.parserCachePurge are *really* slow.

### ISSUE THREAD

**Issue by Nikerabbit** [SemanticMediaWiki/SemanticMediaWiki#3537]
### Setup and configuration

- SMW version: 3.0.0
- MW version: 1.31.1 (release branc)
- PHP version: 7.1.23 (fpm-fcgi)
- DB system (MySQL, Blazegraph, etc.) and version: 10.1.37-MariaDB

### Issue

Some smw.parserCachePurge are *really* slow. Here is an example from my jobqueue log:
```
marras 14 15:51:49 sanat2.csc.fi php[18888]: 2018-11-14 13:51:49 smw.parserCachePurge Nimitys:atlant idlist=[1126576,515,522] exec.mode=exec.journal requestId=W@v1JKL0QgGTGh21LgdIuwAAABA (id=20238,timestamp=20181114101256) STARTING
marras 14 15:55:59 sanat2.csc.fi php[18888]: 2018-11-14 13:55:59 smw.parserCachePurge Nimitys:atlant idlist=[1126576,515,522] exec.mode=exec.journal requestId=W@v1JKL0QgGTGh21LgdIuwAAABA (id=20238,timestamp=20181114101256) t=249336 good
marras 14 15:56:17 sanat2.csc.fi php[18888]: 2018-11-14 13:56:17 smw.parserCachePurge Taidehistoria:akvedukti idlist=array(21) exec.mode=exec.journal requestId=W@vW0YeA1MLTjufzNCeSRQAAAAs (id=20093,timestamp=20181114080333) STARTING
marras 14 16:55:01 sanat2.csc.fi php[18888]: 2018-11-14 14:55:01 smw.parserCachePurge Taidehistoria:akvedukti idlist=array(21) exec.mode=exec.journal requestId=W@vW0YeA1MLTjufzNCeSRQAAAAs (id=20093,timestamp=20181114080333) t=3524463 good
marras 14 16:55:01 sanat2.csc.fi php[18888]: 2018-11-14 14:55:01 smw.parserCachePurge Taidehistoria:amfora idlist=[541] exec.mode=exec.journal requestId=W@vfvEPCI8Kytjf1CL9pygAAABs (id=20131,timestamp=20181114084134) STARTING
marras 14 16:55:01 sanat2.csc.fi php[18888]: 2018-11-14 14:55:01 smw.parserCachePurge Taidehistoria:amfora idlist=[541] exec.mode=exec.journal requestId=W@vfvEPCI8Kytjf1CL9pygAAABs (id=20131,timestamp=20181114084134) t=21 good
marras 14 16:55:01 sanat2.csc.fi php[18888]: 2018-11-14 14:55:01 smw.parserCachePurge Nimitys:still_life idlist=array(17) exec.mode=exec.journal requestId=W@vrh9XGjbbzD2@ByoxlKgAAAAw (id=20176,timestamp=20181114093155) STARTING
marras 14 17:03:23 sanat2.csc.fi php[18888]: 2018-11-14 15:03:23 smw.parserCachePurge Nimitys:still_life idlist=array(17) exec.mode=exec.journal requestId=W@vrh9XGjbbzD2@ByoxlKgAAAAw (id=20176,timestamp=20181114093155) t=501239 good
```

These are in milliseconds, so for example one job took 3524463 ms ~1 hour. This is unacceptable, because it blocks the jobqueue and many jobs will be delayed, and it is not able to keep up with the jobs. I don't have debugged yet why some requests take a long time and why others are fast.

From looking at mytop, I gather it is executing queries like below sequentially, each one taking about 10 seconds:
```
EXPLAIN SELECT /* SMW\SQLStore\QueryDependency\QueryDependencyLinksStore::findDependencyTargetLinks apache@sanat2.c... */ DISTINCT s_id FROM `smw_query_links` WHERE o_id IN ('504','515','1126558','581','557','1126559','1126560','1126561','1126562','530','1126563','558','569','546','555','541','543','595760','1126380','680','681','525','376573','526','516','671','1722','529','1126564','531','512','1126565','565','224971','1126566','1126567','1126568','535','1126551','1126553','1126554','1126555','1126556','1126557','33','610','376583','570927','1123291','4','30','32','31','37','36','35','34','568','578') LIMIT 250200,301:

*** row 1 ***
          table:  smw_query_links
           type:  range
  possible_keys:  o_id
            key:  o_id
        key_len:  4
            ref:  NULL
           rows:  1836043
          Extra:  Using index condition; Using temporary
```

### Steps to reproduce
Not yet clear, but having huge sme_query_links table seems to be a part of it.

**Comment by kghbln** [comment]
Thanks a lot for reporting. Indeed the job queue is a constant source of pain not just for Semantic MediaWiki. I guess having a complex template with many cross-references is at play.

**Comment by Nikerabbit** [comment]
I believe happens with `$smwgEnabledQueryDependencyLinksStore = true;` and goes away if I disable it again and also happens when editing pages (in the command line at least).

**Comment by Nikerabbit** [comment]
And the site is https://tieteentermipankki.fi in case anyone is interested looking at it. It is certainly using a lot of queries, so I am not surprised it puts pressure on query tracking.

**Comment by mwjames** [comment]
> Some smw.parserCachePurge are really slow. 
> And the site is https://tieteentermipankki.fi in case anyone is interested looking at it. It is certainly using a lot of queries, so I am not surprised it puts pressure on query tracking. 

I knew from past observations that those use cases would appear and the job/post-processing via ` smw.parserCachePurge` would hardly be scalable (supported by above results).

I came up with a different approach but requires a new DB field to monitor certain information to avoid job related updates (or other post-processing related tasks) hence it will only be available with SMW 3.1.
